xtm/xtm_api.cc

```diff
@@ -0,0 +1,282 @@
+ * @param[in] num - count of elements to get
+ * @retval return the number of data pointers actually read.
+ */
+static inline unsigned
+xtm_queue_get_data(struct xtm_queue *xtm_queue,
+		   void **data, unsigned num)
+{
+	struct xtm_scsp_queue<union xtm_msg> *queue = &xtm_queue->queue;
+	unsigned i = 0;
+	unsigned new_read = queue->read;
+	unsigned queue_write =
+		__atomic_load_n(&queue->write, __ATOMIC_ACQUIRE);
+
+	for (i = 0; i < num; i++) {
+		if (new_read == queue_write)
+			break;
+		data[i] = queue->buffer[new_read].data;
+		new_read = (new_read + 1) & (queue->len - 1);
+	}
+	__atomic_store_n(&queue->read, new_read, __ATOMIC_RELEASE);
+	return i;
+}
+
+/**
+ * Function executes all functions contained in all queue elements.
+ * Used in case dispatch/invoke xtm API pattern
+ * @param[in] queue - xtm queue
+ * @retval return count of executed functions
+ */
+static inline unsigned
+xtm_queue_execute_funs(struct xtm_queue *xtm_queue)
+{
+	struct xtm_scsp_queue<union xtm_msg> *queue = &xtm_queue->queue;
+	unsigned i = 0;
+	unsigned new_read = queue->read;
+	unsigned queue_write =
+		__atomic_load_n(&queue->write, __ATOMIC_ACQUIRE);
+
+	for(;;) {
+		if (new_read == queue_write)
+			break;
+		queue->buffer[new_read].fun(queue->buffer[new_read].fun_arg);
+		new_read = (new_read + 1) & (queue->len - 1);
+		i++;
+	}
+	__atomic_store_n(&queue->read, new_read, __ATOMIC_RELEASE);
+	return i;
+}
+
+static inline int
+xtm_queue_flush_pipe(struct xtm_queue *xtm_queue)
+{
+	unsigned char tmp[XTM_PIPE_SIZE];
+	ssize_t read_bytes;
+	while ((read_bytes = read(xtm_queue->readfd, tmp, sizeof(tmp))) < 0 &&
+	        errno == EINTR)
+		;
+	if (read_bytes <= 0 && errno != EAGAIN)
+		return -1;
+	return 0;
+}
+
+
+struct xtm_queue *
+xtm_create(unsigned size)
+{
+	int filedes[2];
+	struct xtm_queue *queue = (struct xtm_queue *)
+		malloc(sizeof(struct xtm_queue) +
+		       size * sizeof(union xtm_msg));
+	if (queue == NULL)
+		return NULL;
+
+#ifdef TARANTOOL_XTM_USE_EVENTFD
+	if ((filedes[0] = eventfd(0, 0)) < 0)
+#else
+	if (pipe(filedes) < 0)
+#endif
+		goto free_queue;
+
+#ifdef TARANTOOL_XTM_USE_EVENTFD
+	queue->readfd = queue->writefd = filedes[0];
+#else
+	queue->readfd = filedes[0];
+	queue->writefd = filedes[1];
+#endif
+
+	if (fcntl(queue->readfd, F_SETFL, O_NONBLOCK) < 0 ||
+	    fcntl(queue->writefd, F_SETFL, O_NONBLOCK) < 0)
+		goto close_fd;
+	if (xtm_scsp_queue_init(&queue->queue, size) < 0) {
+		errno = EINVAL;
+		goto close_fd;
+	}
+	return queue;
+
+close_fd:
+	close(queue->readfd);
+	if (queue->readfd != queue->writefd)
+		close(queue->writefd);
+free_queue:
+	free(queue);
+	return NULL;
+};
+
+int
+xtm_delete(struct xtm_queue *queue)
+{
+	int rc = 0;
+	if (close(queue->readfd) < 0)
+		rc = -1;
+	if (queue->readfd != queue->writefd && close(queue->writefd) < 0)
+		rc = -1;
+	free(queue);
+	return rc;
+}
+
+int
+xtm_msg_probe(struct xtm_queue *queue)
+{
+	if (xtm_scsp_queue_free_count(&queue->queue) == 0) {
+		errno = ENOBUFS;
```

Comment:
I don't think you need an error here. The function works like a flag, it does not do anything. Only checks if there is free space. A check for that can't really fail. It can either say "there is space" or "there is no space". Not "I couldn't check if there is space". That means it must return a bool, do not set an error, and be called not probe. Probe assumes a change, an attempt to do something, while here you just check.

Reply:
I talk about this with @Mons. He provided this API and recommended not to touch it

Reply:
Hm, well. I am not buying it, I am really sorry. What is the real reason we need an error here? "Мама сказала в бидончике" does not count.

Reply:
> Hm, well. I am not buying it, I am really sorry. What is the real reason we need an error here? "Мама сказала в бидончике" does not count.

That's what I said... again.